### PR TITLE
[SYCL] Check if reqd_work_group_size dim is valid

### PR DESCRIPTION
**Pull request by KornevNikita**
The number of arguments must match the dimensionality of the work-group used to invoke the kernel.

### CODE REVIEW

**Comment by KornevNikita** [comment]
@intel/llvm-reviewers-runtime could you take a look please

**Comment by maarquitos14** [comment]
> The number of arguments must match the dimensionality of the work-group used to invoke the kernel. The spec doesn't require the implementation to throw the exception, but it should improve user experience.

From the spec: 

> If the kernel is submitted to a device that does support the work-group size, but the application provides an nd_range that does not match the size from the attribute, then the implementation must throw a synchronous exception with the errc::nd_range error code.

Isn't this implicitly saying that we should throw? It seems to me that not matching the dimensionality is implicitly included in not matching the size from the attribute.

**Comment by KornevNikita** [comment]
> If the kernel is submitted to a device that does support the work-group size, but the application provides an nd_range that does not match the size from the attribute, then the implementation must throw a synchronous exception with the errc::nd_range error code.

I thought the spec demands attribute and nd_range to have the same dimensionality otherwise it's a user's problem if this doesn't work. But yep, it seems to me like we can interpret the spec this way:

> matching the dimensionality is implicitly included in not matching the size from the attribute.

Updated the commit message.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/test-e2e/Basic/reqd_work_group_size_matches_dimensions.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%{sycl_triple}  %s -o %t.out
+// RUN: %{run} %t.out
+
+// UNSUPPORTED: hip
+
+#include "sycl/sycl.hpp"
+
+using namespace sycl;
+
+#define CHECK_INVALID_REQD_WORK_GROUP_SIZE(Range, Item, ...)                   \
```
[the remark this answers is not in the document]
Comment (KornevNikita) [reply sycl/test-e2e/Basic/reqd_work_group_size_matches_dimensions.cpp:10]: Compiler says `attribute 'reqd_work_group_size' does not support argument pack expansion [[sycl::reqd_work_group_size(args...)]] {});`

File: sycl/test-e2e/Basic/reqd_work_group_size_matches_dimensions.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%{sycl_triple}  %s -o %t.out
+// RUN: %{run} %t.out
+
+// UNSUPPORTED: hip
```
[the remark this answers is not in the document]
Comment (KornevNikita) [reply sycl/test-e2e/Basic/reqd_work_group_size_matches_dimensions.cpp:5]: Added this link in b703085: https://github.com/intel/llvm/issues/9353
It's also turned off in https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Basic/reqd_work_group_size.cpp#L4
I assume that wrong case can't be compiled for hip.

File: sycl/test-e2e/Basic/reqd_work_group_size_matches_dimensions.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%{sycl_triple}  %s -o %t.out
```
[the remark this answers is not in the document]
Comment (KornevNikita) [reply sycl/test-e2e/Basic/reqd_work_group_size_matches_dimensions.cpp]: Fixed [b703085](https://github.com/intel/llvm/pull/12179/commits/b7030852c74d9e9f3d92ccbb083da2d111165280)

File: sycl/test-e2e/Scheduler/HandleException.cpp [hunk 4]
```diff
@@ -3,6 +3,13 @@
 
 // Expected exception is generated for OpenCL backend only.
 // REQUIRES: opencl
+
+// The number of arguments of "reqd_work_group_size" must match the
+// dimensionality of the work-group used to invoke the kernel.
+// But "cl::reqd_work_group_size" always accepts 3 arguments.
+// Not sure if this test should exist.
+// XFAIL: *
```
[the remark this answers is not in the document]
Comment (KornevNikita) [reply sycl/test-e2e/Scheduler/HandleException.cpp]: Removed [b703085](https://github.com/intel/llvm/pull/12179/commits/b7030852c74d9e9f3d92ccbb083da2d111165280)

**Review by aelovikov-intel — APPROVED** [review]

**Comment by KornevNikita** [comment]
@intel/llvm-gatekeepers could you merge this one please.

**Pull request merged**